### PR TITLE
pkg/nixpath: Reduce allocations when validating

### PR DESCRIPTION
**Pull request by adisbladis**
This is a follow-up to https://github.com/nix-community/go-nix/pull/77.

This ensures we don't have to do costly allocations for every store path that we validate when validating a derivation.

The following test program:
``` go
package main

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"testing"
	"time"

	"github.com/nix-community/go-nix/pkg/derivation"
)

func main() {

	drvPath := "/nix/store/lfb109iifgph2xa94aps13n2w8bpbqsq-texlive-combined-full-2021-final.drv"

	n := 50

	derivationFile, err := os.Open(drvPath)
	if err != nil {
		panic(err)
	}

	drvBytes, err := io.ReadAll(derivationFile)
	if err != nil {
		panic(err)
	}

	{
		start := time.Now()

		for i := 0; i < n; i++ {
			_, err := derivation.ReadDerivation(bytes.NewReader(drvBytes))
			if err != nil {
				panic(err)
			}
		}

		fmt.Println(time.Since(start) / time.Duration(n))
	}

	fmt.Println("Allocs:", int(testing.AllocsPerRun(n, func() {
		_, err := derivation.ReadDerivation(bytes.NewReader(drvBytes))
		if err != nil {
			panic(err)
		}
	})))

}
```

Yields the following numbers:
- Before
```
14.045441ms
Allocs: 27304
```

- After
```
8.503212ms
Allocs: 11864
```

### CODE REVIEW

File: pkg/nixbase32/nixbase32.go [hunk 1]
```diff
@@ -8,6 +8,42 @@ import (
 // Alphabet contains the list of valid characters for the Nix base32 alphabet.
 const Alphabet = "0123456789abcdfghijklmnpqrsvwxyz"
 
+// ValidateString validates if a byte slice is valid nixbase32.
+func Validate(b []byte) error {
```
Comment (flokli) [inline pkg/nixbase32/nixbase32.go:12]: The comment still stays `ValidateString`, but it's accepting bytes. Maybe we should call them `Validate{String,Bytes}`?

File: pkg/nixbase32/nixbase32.go [hunk 2]
```diff
@@ -8,6 +8,42 @@ import (
...
+		carry := byte(digit) >> (8 - j)
+
+		// if we're at the end of dst…
+		if i == uint(dstLen-1) {
+			// but have a nonzero carry, the encoding is invalid.
+			if carry != 0 {
+				return fmt.Errorf("invalid encoding")
+			}
+		}
+	}
```
Comment (flokli) [inline pkg/nixbase32/nixbase32.go:42]: This is pretty much copy-pasted from `DecodeString`. Can we move this into a private function and use it from both places?

Comment (adisbladis) [reply pkg/nixbase32/nixbase32.go]: Done.

File: pkg/nixbase32/nixbase32.go (hunk 1 above)
Comment (adisbladis) [reply pkg/nixbase32/nixbase32.go]: Done.

**Pull request merged**